platform-sdk/swirlds-common/src/main/java/com/swirlds/common/wiring/builders/TaskSchedulerType.java

```diff
@@ -0,0 +1,74 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.swirlds.common.wiring.builders;
+
+/**
+ * Various types of task schedulers. Pass one of these types to {@link TaskSchedulerBuilder#withType(TaskSchedulerType)}
+ * to create a task scheduler of the desired type. If unspecified, the default scheduler type is {@link #SEQUENTIAL}.
+ */
+public enum TaskSchedulerType {
+    /**
+     * Tasks are executed in a fork join pool one at a time in the order they were enqueued. There is a happens before
+     * relationship between each task.
+     */
+    SEQUENTIAL,
+    /**
+     * Tasks are executed on a dedicated thread one at a time in the order they were enqueued. There is a happens before
+     * relationship between each task. This scheduler type has very similar semantics as {@link #SEQUENTIAL}, although
+     * the implementation and performance characteristics are not identical.
+     */
+    SEQUENTIAL_THREAD,
+    /**
+     * Tasks are executed on a fork join pool. Tasks may be executed in parallel with each other. Ordering is not
+     * guaranteed.
+     */
+    CONCURRENT,
+    /**
+     * Tasks are executed immediately on the callers thread. There is no queue for tasks waiting to be handled (logical
+     * or otherwise). Useful for scenarios where tasks are extremely small and not worth the scheduling overhead.
+     * <p>
+     * Only a single logical thread of execution is permitted to send data to a direct task scheduler.
+     * {@link #SEQUENTIAL} and {@link #SEQUENTIAL_THREAD} are permitted to send data to a direct task scheduler, but it
+     * is illegal for more than one of these schedulers to send data to the same direct task scheduler. It is legal for
+     * operations that are executed on the calling thread (e.g. filters, transformers, other direct executors) to call
+     * into a direct executor. Such calls are treated as equivalent to the parent sequential scheduler calling into the
+     * direct scheduler.
+     *
+     * <p>
+     * {@link #CONCURRENT} task schedulers are forbidden from sending data to a direct task scheduler.
+     *
+     * <p>
+     * These constraints are enforced by the framework.
+     */
+    DIRECT,
```

Comment:
maybe we should call this one `DIRECT_SEQUENTIAL` since the other one has a `STATELESS` suffix

Reply:
I'm not sure this name fits. It's legal to call a `DIRECT` scheduler from other scheduler types. Legal types are

- `SEQUENTIAL`
- `SEQUENTIAL_THREAD`
- `DIRECT`
- `DIRECT_STATELESS`
- transformers
- splitters
- filters
- the special type of transformer for state reservations (once I write it)

I've updated the javadoc to be more precise.

```
    /**
     * Tasks are executed immediately on the caller's thread. There is no queue for tasks waiting to be handled (logical
     * or otherwise). Useful for scenarios where tasks are extremely small and not worth the scheduling overhead.
     * <p>
     * Only a single logical thread of execution is permitted to send data to a direct task scheduler.
     * {@link #SEQUENTIAL} and {@link #SEQUENTIAL_THREAD} schedulers are permitted to send data to a direct task
     * scheduler, but it is illegal for more than one of these schedulers to send data to the same direct task
     * scheduler. {@link #CONCURRENT} task schedulers are forbidden from sending data to a direct task scheduler. It is
     * legal for operations that are executed on the calling thread (e.g. filters, transformers, stateless/stateful
     * direct schedulers) to call into a direct scheduler as long as the calling thread is not in a concurrent scheduler
     * or originating from more than one sequential scheduler.
     * <p>
     * To decide if a direct scheduler is wired in a legal way, the following algorithm is used:
     * <ul>
     * <li>Create a directed graph where vertices are schedulers and edges are wires between schedulers</li>
     * <li>Starting from each vertex, walk over the graph in depth first order. Follow edges that lead to
     * DIRECT or DIRECT_STATELESS vertices, but do not follow edges that lead into SEQUENTIAL, SEQUENTIAL_THREAD,
     * or CONCURRENT vertices.</li>
     * <li>If a DIRECT vertex is reachable starting from a CONCURRENT vertex, the wiring is illegal.</li>
     * <li>For each vertex with type DIRECT, count the number of unique SEQUENTIAL or SEQUENTIAL_THREAD vertexes that
     * it can be reached by. If that number exceeds 1, then the wiring is illegal.</li>
     * </ul>
     *
     * <p>
     * These constraints are enforced by the framework.
     */
```

Reply:
I talked with Lazar, he is ok with the current naming.